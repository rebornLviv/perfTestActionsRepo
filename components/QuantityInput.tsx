import React from "react";
import { StyleSheet, View, Text } from "react-native";

import { IconButton } from "./IconButton";

export type QuantityInputProps = {
  value: number;
  onIncrease: () => void;
  onDecrease: () => void;
  minValue?: number;
  maxValue?: number;
};

export const QuantityInput = ({
  value,
  onDecrease,
  onIncrease,
  minValue = 0,
  maxValue = 100,
}: QuantityInputProps) => {
  const handleDecrease = () => {
    if (value > minValue) {
      onDecrease();
    }
  };
  const handleIncrease = () => {
    if (value < maxValue) {
      onIncrease();
    }
  };

  return (
    <View style={[styles.container, styles.quantityContainer]}>
      <IconButton
        onPress={handleDecrease}
        disabled={value === minValue}
        accessibilityLabel="Decrease quantity button"
        name="minus-circle"
      />
      <Text
        style={[styles.quantity]}
        accessibilityLabel={`Quantity: ${value}`}
        accessibilityValue={{ min: minValue, now: value, max: maxValue }}
      >
        {value}
      </Text>
      <IconButton
        onPress={handleIncrease}
        disabled={value === maxValue}
        accessibilityLabel="Increase quantity button"
        name="plus-circle"
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
  },
  quantityContainer: {
    borderRadius: 16,
  },
  quantity: {
    paddingHorizontal: 8,
    textAlign: "center",
    minWidth: 40,
    height: 40,
    lineHeight: 40,
  },
});
