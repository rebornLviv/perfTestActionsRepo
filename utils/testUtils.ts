import { userEvent } from "@testing-library/react-native";

export const rntlUser = userEvent.setup();
