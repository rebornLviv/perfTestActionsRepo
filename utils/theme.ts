import { DefaultTheme } from "@react-navigation/native";

export const theme = {
  ...DefaultTheme,
  colors: {
    ...DefaultTheme.colors,
    primary: "#8232ff",
  },
};
