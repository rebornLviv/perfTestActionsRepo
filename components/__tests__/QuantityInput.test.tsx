import { render, screen } from "@testing-library/react-native";
import React, { useState } from "react";

import { rntlUser } from "../../utils/testUtils";
import { QuantityInput } from "../QuantityInput";

describe("<QuantityInput />", function () {
  it("Renders properly", () => {
    render(
      <QuantityInput value={667} onIncrease={() => {}} onDecrease={() => {}} />,
    );

    expect(screen.getByText("667")).toBeOnTheScreen();
  });

  it("Renders properly with snapshot", () => {
    const view = render(
      <QuantityInput value={667} onIncrease={() => {}} onDecrease={() => {}} />,
    );

    expect(view).toMatchSnapshot();
  });

  describe("Testing with mocked functions, NOT RECOMMENDED 👎", () => {
    it("Increase and decrease quantity are called properly", async () => {
      const onIncreaseMocked = jest.fn();
      const onDecreaseMocked = jest.fn();

      render(
        <QuantityInput
          value={10}
          onIncrease={onIncreaseMocked}
          onDecrease={onDecreaseMocked}
        />,
      );

      await rntlUser.press(screen.getByLabelText("Increase quantity button"));
      await rntlUser.press(screen.getByLabelText("Decrease quantity button"));

      expect(onIncreaseMocked).toHaveBeenCalled();
      expect(onDecreaseMocked).toHaveBeenCalled();
    });

    it("Increase and decrease quantity are not called in case when of min & max values", async () => {
      const onIncreaseMocked = jest.fn();
      const onDecreaseMocked = jest.fn();

      render(
        <QuantityInput
          value={10}
          minValue={10}
          maxValue={10}
          onIncrease={onIncreaseMocked}
          onDecrease={onDecreaseMocked}
        />,
      );

      await rntlUser.press(screen.getByLabelText("Increase quantity button"));
      await rntlUser.press(screen.getByLabelText("Decrease quantity button"));

      expect(onIncreaseMocked).not.toHaveBeenCalled();
      expect(onDecreaseMocked).not.toHaveBeenCalled();
    });
  });

  describe("Testing with usage example, RECOMMENDED 👍", () => {
    // Example component with usage of quantity input
    const QuantityInputWrapper = () => {
      const [value, setValue] = useState(0);

      return (
        <QuantityInput
          value={value}
          onIncrease={() => setValue(value + 1)}
          onDecrease={() => setValue(value - 1)}
          minValue={0}
          maxValue={2}
        />
      );
    };

    it("Increase and decrease quantity are called properly", async () => {
      render(<QuantityInputWrapper />);

      // initial state
      expect(screen.getByText("0")).toBeOnTheScreen();

      // Added one quantity and check result
      await rntlUser.press(screen.getByLabelText("Increase quantity button"));
      expect(screen.getByText("1")).toBeOnTheScreen();

      // Subtracted one quantity and check result
      await rntlUser.press(screen.getByLabelText("Decrease quantity button"));
      expect(screen.getByText("0")).toBeOnTheScreen();
    });

    it("Increase and decrease quantity are not called in case when of min & max values", async () => {
      render(<QuantityInputWrapper />);

      const decreaseButton = screen.getByLabelText("Decrease quantity button");
      const increaseButton = screen.getByLabelText("Increase quantity button");

      // initial state
      expect(screen.getByText("0")).toBeOnTheScreen();

      // Subtracted one quantity and check result tah shouldn't be lower than minimal value
      await rntlUser.press(decreaseButton);
      await rntlUser.press(decreaseButton);

      expect(screen.getByText("0")).toBeOnTheScreen();

      // Added one quantity and check result that shouldn't be greater than max value
      await rntlUser.press(increaseButton);
      await rntlUser.press(increaseButton);
      await rntlUser.press(increaseButton);
      expect(screen.getByText("2")).toBeOnTheScreen();
    });
  });
});
