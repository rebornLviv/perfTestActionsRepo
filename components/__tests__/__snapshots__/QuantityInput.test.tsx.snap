// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuantityInput /> Renders properly with snapshot 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      },
      {
        "borderRadius": 16,
      },
    ]
  }
>
  <View
    accessibilityLabel="Decrease quantity button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        false,
        false,
        undefined,
      ]
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": undefined,
            "fontSize": 32,
          },
          undefined,
          {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      
    </Text>
  </View>
  <Text
    accessibilityLabel="Quantity: 667"
    accessibilityValue={
      {
        "max": 100,
        "min": 0,
        "now": 667,
      }
    }
    style={
      [
        {
          "height": 40,
          "lineHeight": 40,
          "minWidth": 40,
          "paddingHorizontal": 8,
          "textAlign": "center",
        },
      ]
    }
  >
    667
  </Text>
  <View
    accessibilityLabel="Increase quantity button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        false,
        false,
        undefined,
      ]
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": undefined,
            "fontSize": 32,
          },
          undefined,
          {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;
